### DISCUSSION TITLE
Is the shutdown event supposed to run when uvicorn is set to --reload?

### DISCUSSION THREAD

**Discussion by camerondavison** [fastapi/fastapi#9112] [Questions] [answered]
### Example

```Python
from fastapi import FastAPI

app = FastAPI()


@app.get("/")
def read_root():
    return {"Hello": "World"}


@app.on_event("startup")
def startup():
    print('startup')


@app.on_event("shutdown")
def shutdown():
    print('shutdown')

```

### Environment

* OS: [e.g. Linux / Windows / macOS]: mac
* FastAPI Version [e.g. 0.3.0]: 0.61.0
* Python: 3.7.7

### Additional context

```
❯ uvicorn example:app
INFO:     Started server process [17611]
INFO:     Waiting for application startup.
startup
INFO:     Application startup complete.
INFO:     Uvicorn running on http://127.0.0.1:8000 (Press CTRL+C to quit)
^CINFO:     Shutting down
INFO:     Waiting for application shutdown.
shutdown
INFO:     Application shutdown complete.
INFO:     Finished server process [17611]

❯ uvicorn example:app --reload
INFO:     Uvicorn running on http://127.0.0.1:8000 (Press CTRL+C to quit)
INFO:     Started reloader process [17670] using statreload
INFO:     Started server process [17690]
INFO:     Waiting for application startup.
startup
INFO:     Application startup complete.
^CINFO:     Shutting down
INFO:     Finished server process [17690]
INFO:     Stopping reloader process [17670]
```

**Comment by ycd** [comment]
When we run uvicorn with `--reload` flag, two threads will be started, one with the server, one to monitor file changes, when a file change is detected it restarts the dev server.

When you start a server, it's just a process, you can see the PID within the square brackets

```python
def restart(self):
    self.mtimes = {}
    os.kill(self.process.pid, signal.SIGTERM)
    self.process.join()

    self.process = get_subprocess(
        config=self.config, target=self.target, sockets=self.sockets
    )
    self.process.start()

```

You can see how Uvicorn restarts, it kills the currenct process than starts a new one, so to answer to your question,  it is supposed to run.

**Comment by codethief** [comment]
@ycd Thanks so much for the insight! I arrived here because I was wondering about the exact same thing. Maybe your explanation could be added to the docs? (E.g. [here](https://fastapi.tiangolo.com/advanced/events/)) In my case, I need to fire up a background process right when the uvicorn server starts and make sure to kill it when the uvicorn server stops, so learning about the life cycle of the server process was crucial.

**Comment by camerondavison** [comment]
given
> so to answer to your question, it is supposed to run.

and it is not, is this a bug?

**Comment by PhilippHomann** [comment]
Just stumpled on that too..
Maybe there should be a note in the docs that this will not work with _--reload_?
Or a warning if there is a shutdown event registered, but _--reload_ is used?

Don't know whether this is in scope of fastapi or of uvicorn?!

**Comment by YuriiMotov** [comment, marked as answer]
Works fine in current versions (uvicorn 0.35.0, fastapi 0.115.14, starlette 0.46.2):

```
INFO:     Will watch for changes in these directories: ['/home/yurii/code/fast_tests']
INFO:     Uvicorn running on http://127.0.0.1:8000 (Press CTRL+C to quit)
INFO:     Started reloader process [18442] using WatchFiles
INFO:     Started server process [18444]
INFO:     Waiting for application startup.
startup
INFO:     Application startup complete.
WARNING:  WatchFiles detected changes in 'tt.py'. Reloading...
INFO:     Shutting down
INFO:     Waiting for application shutdown.
shutdown
INFO:     Application shutdown complete.
INFO:     Finished server process [18444]
INFO:     Started server process [18492]
INFO:     Waiting for application startup.
startup
INFO:     Application startup complete.
^CINFO:     Shutting down
INFO:     Waiting for application shutdown.
shutdown
INFO:     Application shutdown complete.
INFO:     Finished server process [18492]
INFO:     Stopping reloader process [18442]
```
